caas/kubernetes/provider/k8swatcher.go

```diff
@@ -46,43 +48,55 @@ func newKubernetesNotifyWatcher(wi watch.Interface, name string, clock jujuclock
 
 const sendDelay = 1 * time.Second
 
+func (w *kubernetesNotifyWatcher) OnAdd(obj interface{}) {
+	logger.Tracef("received k8s event: Added")
+	if pod, ok := obj.(*core.Pod); ok {
+		logger.Tracef("%v(%v) = %v, status=%+v", pod.Name, pod.UID, pod.Labels, pod.Status)
+	}
+	if ns, ok := obj.(*core.Namespace); ok {
+		logger.Tracef("%v(%v) = %v, status=%+v", ns.Name, ns.UID, ns.Labels, ns.Status)
+	}
+
+	w.out <- struct{}{}
+	logger.Debugf("fire notify watcher for %v", w.name)
+}
+
+func (w *kubernetesNotifyWatcher) OnDelete(obj interface{}) {
+	logger.Tracef("received k8s event: Deleted")
+	if pod, ok := obj.(*core.Pod); ok {
+		logger.Tracef("%v(%v) = %v, status=%+v", pod.Name, pod.UID, pod.Labels, pod.Status)
+	}
+	if ns, ok := obj.(*core.Namespace); ok {
+		logger.Tracef("%v(%v) = %v, status=%+v", ns.Name, ns.UID, ns.Labels, ns.Status)
+	}
+
+	w.out <- struct{}{}
+	logger.Debugf("fire notify watcher for %v", w.name)
+}
+
+func (w *kubernetesNotifyWatcher) OnUpdate(oldObj, newObj interface{}) {
+	logger.Tracef("received k8s event: Updated")
+
+	if pod, ok := newObj.(*core.Pod); ok {
+		logger.Tracef("%v(%v) = %v, status=%+v", pod.Name, pod.UID, pod.Labels, pod.Status)
+	}
+	if ns, ok := newObj.(*core.Namespace); ok {
+		logger.Tracef("%v(%v) = %v, status=%+v", ns.Name, ns.UID, ns.Labels, ns.Status)
+	}
+	w.out <- struct{}{}
+	logger.Debugf("fire notify watcher for %v", w.name)
+}
+
 func (w *kubernetesNotifyWatcher) loop() error {
 	defer close(w.out)
-	defer w.k8watcher.Stop()
 
-	// Set out now so that initial event is sent.
-	out := w.out
-	var delayCh <-chan time.Time
+	informerStopCh := make(chan struct{})
+	go w.informer.Run(informerStopCh)
 
-	for {
-		select {
-		case <-w.catacomb.Dying():
-			return w.catacomb.ErrDying()
-		case evt, ok := <-w.k8watcher.ResultChan():
```

Comment:
This old logic is for two things, collated events with a delay and triggering an initial event. It also makes sure we don't wait for the consumer.